src/serializer/Referentializer.js

```diff
@@ -194,8 +194,15 @@ export class Referentializer {
 
           // Already referentialized in prior scope
           if (residualBinding.declarativeEnvironmentRecord) {
-            invariant(residualBinding.scope);
-            instance.scopeInstances.add(residualBinding.scope);
+            const scope = residualBinding.scope;
+            invariant(scope);
+            let exists = false;
+            for (const s of instance.scopeInstances) {
+              if (s.name === scope.name) exists = true;
```

Comment:
You could just return here. That terminates the search and avoids the add, without the need of an extra local variable.

Reply:
mhh not sure if that's a good idea
what if someone adds stuff after this `if` block?

Reply:
At least `break` out of the for-loop when a match is found.

Reply:
Consider changing scopeInstances from "Set<...>" to "Map<string, ...>". It will improve algorithm perf and much simplify the code.

Reply:
The best way to make sure that a future programmer does not break something is to write very simple, easy to understand code and to add comments. Writing something that seems wrong, impedes understanding and does not help the future programmer.